odbc_fdw.c

```diff
@@ -155,9 +182,208 @@ odbc_fdw_handler(PG_FUNCTION_ARGS)
+	options->table        = NULL;
+	options->prefix       = NULL;
+	options->sql_query    = NULL;
+	options->sql_count    = NULL;
+	options->mapping_list = NIL;
+}
+
+void
+copy_odbcFdwOptions(odbcFdwOptions* to, odbcFdwOptions* from)
+{
+	if (to && from)
+	{
+		to->dsn          = from->dsn;
+		to->driver       = from->driver;
+		to->host         = from->host;
+		to->port         = from->port;
+		to->database     = from->database;
+		to->username     = from->username;
+		to->password     = from->password;
+		to->schema       = from->schema;
+		to->table        = from->table;
+		to->prefix       = from->prefix;
+		to->sql_query    = from->sql_query;
+		to->sql_count    = from->sql_count;
+		to->mapping_list = from->mapping_list;
+	}
+}
+
+/*
+ * Replace empty string by null pointer
+ */
+static void
+normalize_empty_string(char **str)
+{
+	if (*str && !**str)
+	{
+		*str = NULL;
+	}
+}
+
+/*
+ * Avoid NULL string: return original string, or empty string if NULL
+ */
+static char*
+empty_string_if_null(char *string)
+{
+	static char* empty_string = "";
+	return string == NULL ? empty_string : string;
+}
+
+static void
+extract_odbcFdwOptions(List *options_list, odbcFdwOptions *extracted_options)
+{
+	ListCell        *lc;
+	List            *mapping_list;
+
+	#ifdef DEBUG
+		elog(DEBUG1, "extract_init_odbcFdwOptions");
+	#endif
+
+	init_odbcFdwOptions(extracted_options);
+
+	/* Loop through the options, and get the foreign table options */
+	foreach(lc, options_list)
+	{
+		DefElem *def = (DefElem *) lfirst(lc);
+
+		if (strcmp(def->defname, "dsn") == 0)
+		{
+			extracted_options->dsn = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "driver") == 0)
+		{
+			extracted_options->driver = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "host") == 0)
+		{
+			extracted_options->host = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "port") == 0)
+		{
+			extracted_options->port = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "database") == 0)
+		{
+			extracted_options->database = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "schema") == 0)
+		{
+			extracted_options->schema = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "table") == 0)
+		{
+			extracted_options->table = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "prefix") == 0)
+		{
+			extracted_options->prefix = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "sql_query") == 0)
+		{
+			extracted_options->sql_query = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "sql_count") == 0)
+		{
+			extracted_options->sql_count = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "username") == 0)
+		{
+			extracted_options->username = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "password") == 0)
+		{
+			extracted_options->password = defGetString(def);
+			continue;
+		}
```

Comment:
I'm leaving that for when implementing arbitrary ODBC attributes (we'll need to keep an open set of parameter definitions then)